File: skills/JuliusBrussee--caveman/packages/middleware/python/caveman_middleware/strands.py
"""Strands native Model delegate, including the structured-output hook gap."""
from __future__ import annotations

import asyncio
import json
import uuid
import weakref

try:
    from strands import tool
    from strands.models.model import Model
    from strands.plugins import Plugin
    from strands.types.tools import ToolContext
except ModuleNotFoundError as error:
    raise ImportError("Install caveman-middleware[strands] to use the Strands adapter") from error

from caveman_cloud.middleware import Adapter, Candidate, MiddlewareRuntime, Scope
from caveman_cloud.middleware.runtime import RECOVERY_DESCRIPTION, RECOVERY_SCHEMA
from ._native import Attempt, manifest, owner, plain, replace_path
from ._versions import matches_framework
from ._usage import usage

ADAPTER = Adapter("strands", "0.1.0", "1.55.0", "strands-content-v1")


def _scope(source, state):
    scope = source if isinstance(source, Scope) else source(state or {})
    if not isinstance(scope, Scope):
        raise TypeError("Strands scope resolver must return a Caveman Scope")
    return scope


class CavemanModel(Model):
    """Public native model delegate. A model wrapper alone is recovery-free."""
    def __init__(self, model, *, runtime, scope):
        if not isinstance(model, Model):
            raise TypeError("Expected a native Strands Model")
        self.model = model
        self.runtime = runtime.as_async() if isinstance(runtime, MiddlewareRuntime) else runtime
        self.scope, self.registration = scope, None
        self.version_supported = matches_framework(("strands-agents", "1.55", "2"))
        if not self.version_supported and self.runtime.mode != "off":
            self.runtime.decline("unsupported_version")

    @property
    def stateful(self):
        return self.model.stateful

    def update_config(self, **model_config):
        return self.model.update_config(**model_config)

    def get_config(self):
        return self.model.get_config()

    async def count_tokens(self, *args, **kwargs):
        return await self.model.count_tokens(*args, **kwargs)

    async def _prepare(self, messages, tool_specs, system_prompt, *, tool_choice=None, system_prompt_content=None, invocation_state=None, structured=False):
        if owner.get() is not None:
            return messages, None

        def passive(reason):
            return messages, Attempt(self.runtime, None, str(uuid.uuid4()), str(uuid.uuid4()), passive=True, reason=reason, adapter=ADAPTER.id)

        if self.runtime.mode == "off":
            return passive("disabled")
        if not self.version_supported:
            return passive("unsupported_version")
        if self.stateful:
            return passive("opaque_context")
        prefix = system_prompt_content if system_prompt_content is not None else [{"text": system_prompt}] if system_prompt else []
        context = manifest([{"system": prefix}, *messages])
        if context is None:
            return passive("unsupported_shape")
        scope = _scope(self.scope, invocation_state)
        binding, overhead = None, None
        if not structured and tool_choice in (None, {"auto": {}}) and self.registration and self.registration.bound(tool_specs):
            binding = self.runtime.recovery(scope)
            overhead = json.dumps(self.registration.recovery_tool.tool_spec, ensure_ascii=False, separators=(",", ":"))
        names = {p["toolUse"]["toolUseId"]: p["toolUse"]["name"] for m in messages if plain(m) for p in m.get("content", [])
            if plain(p) and plain(p.get("toolUse")) and type(p["toolUse"].get("toolUseId")) is str and type(p["toolUse"].get("name")) is str}
        candidates, paths = [], {}
        for mi, message in enumerate(messages):
            if not plain(message):
                continue
            for bi, block in enumerate(message.get("content", [])):
                result = block.get("toolResult") if plain(block) else None
                if not plain(result) or type(result.get("toolUseId")) is not str or result["toolUseId"] not in names or names[result["toolUseId"]] == "caveman_retrieve" or result.get("status") == "error":
                    continue
                for pi, part in enumerate(result.get("content", [])):
                    if not plain(part) or set(part) != {"text"} or type(part["text"]) is not str:
                        continue
                    key = f"message-{mi}.block-{bi}.part-{pi}"
                    candidates.append(Candidate(key, part["text"], key))
                    paths[key] = (mi, "content", bi, "toolResult", "content", pi, "text")
        config = self.get_config()
        model_id = config.get("model_id") if plain(config) else None
        attempt = Attempt(self.runtime, scope, str(uuid.uuid4()), str(uuid.uuid4()), adapter=ADAPTER.id)
        result = await self.runtime.optimize(scope=scope, adapter=ADAPTER, manifest=context, candidates=candidates, binding=binding,
            recovery_overhead_text=overhead, model={"provider": type(self.model).__name__, "id": model_id, "protocol": "strands"} if isinstance(model_id, str) else None,
            logical_call_id=attempt.logical_call_id, attempt_id=attempt.attempt_id)
        attempt.optimization = result if not result.replacements else None
        if binding and not self.registration.bound(tool_specs):
            attempt.reason = "recovery_unavailable"
            return messages, attempt
        if not all(item["segment_id"] in paths for item in result.replacements):
            attempt.reason = "invalid_plan"
            return messages, attempt
        view = messages
        for replacement in result.replacements:
            view = replace_path(view, paths[replacement["segment_id"]], replacement["text"])
        attempt.optimization = result
        return view, attempt

    async def _observe(self, iterator, attempt):
        if attempt is None:
            async for event in iterator:
                yield event
            return
        attempt.observe("dispatch_intent")
        last = None
        try:
            while True:
                token = owner.set(attempt)
                try:
                    event = await anext(iterator)
                except StopAsyncIteration:
                    attempt.observe("completed", usage(last))
                    return
                finally:
                    owner.reset(token)
                if not attempt.passive and plain(event) and plain(event.get("metadata")):
                    last = event["metadata"].get("usage") or last
                yield event
        except (asyncio.CancelledError, GeneratorExit):
            attempt.observe("cancelled")
            raise
        except BaseException:
            attempt.observe("failed")
            raise
        finally:
            if hasattr(iterator, "aclose"):
                await iterator.aclose()

    async def stream(self, messages, tool_specs=None, system_prompt=None, *, tool_choice=None, system_prompt_content=None, invocation_state=None, cancel_signal=None, **kwargs):
        if cancel_signal is not None and cancel_signal.is_set():
            if owner.get() is None:
                self.runtime.report(None, reason="cancelled", adapter=ADAPTER.id)
            raise asyncio.CancelledError
        view, attempt = await self._prepare(messages, tool_specs, system_prompt, tool_choice=tool_choice,
            system_prompt_content=system_prompt_content, invocation_state=invocation_state)
        if cancel_signal is not None and cancel_signal.is_set():
            if owner.get() is None:
                self.runtime.report(None, reason="cancelled", adapter=ADAPTER.id)
            raise asyncio.CancelledError
        native = self.model.stream(view, tool_specs, system_prompt, tool_choice=tool_choice,
            system_prompt_content=system_prompt_content, invocation_state=invocation_state, cancel_signal=cancel_signal, **kwargs).__aiter__()
        async for event in self._observe(native, attempt):
            yield event

    async def structured_output(self, output_model, prompt, system_prompt=None, **kwargs):
        view, attempt = await self._prepare(prompt, None, system_prompt, invocation_state=kwargs.get("invocation_state"), structured=True)
        native = self.model.structured_output(output_model, view, system_prompt, **kwargs).__aiter__()
        async for event in self._observe(native, attempt):
            yield event


class _Registration(Plugin):
    name = "caveman:middleware"

    def __init__(self, model):
        self.model, self.agent = model, None
        super().__init__()

        @tool(name="caveman_retrieve", description=RECOVERY_DESCRIPTION, inputSchema={"json": RECOVERY_SCHEMA}, context=True)
        async def recover(handle: str, tool_context: ToolContext, offset: int = 0, limit: int = 262144, query: str = ""):
            if tool_context.cancel_signal.is_set():
                raise asyncio.CancelledError
            scope = _scope(model.scope, tool_context.invocation_state)
            result = await model.runtime.retrieve(scope, handle=handle, offset=offset, limit=limit, query=query)
            return json.dumps(result, ensure_ascii=False, separators=(",", ":"))
        self.recovery_tool = recover

    def init_agent(self, agent):
        if self.agent is not None and self.agent() is not agent:
            raise ValueError("Create a separate Caveman Strands bundle for each agent")
        self.agent = weakref.ref(agent)

    def bound(self, specs):
        agent = self.agent() if self.agent else None
        return (agent is not None and agent.tool_registry.registry.get("caveman_retrieve") is self.recovery_tool
                and type(specs) is list and sum(s.get("name") == "caveman_retrieve" for s in specs if plain(s)) == 1
                and any(s == self.recovery_tool.tool_spec for s in specs))


def with_caveman_model(model, *, runtime, scope):
    return CavemanModel(model, runtime=runtime, scope=scope)


def with_caveman_agent(options: dict, *, runtime, scope) -> dict:
    """Return native Agent constructor options; Strands keeps its own loop."""
    model = CavemanModel(options["model"], runtime=runtime, scope=scope)
    if runtime.mode == "off" or not model.version_supported:
        return {**options, "model": model}
    registration = _Registration(model)
    model.registration = registration
    tools = list(options.get("tools", []))
    if runtime.mode == "compress" and not any(getattr(t, "tool_name", None) == "caveman_retrieve" for t in tools):
        tools.append(registration.recovery_tool)
    return {**options, "model": model, "tools": tools, "plugins": [*options.get("plugins", []), registration]}

File: skills/JuliusBrussee--caveman/packages/middleware/typescript/src/strands.ts
import { Model, Message, TextBlock, ToolResultBlock, FunctionTool, type AgentConfig, type BaseModelConfig,
  type StreamOptions, type CountTokensOptions, type ModelStreamEvent, type LocalAgent, type Plugin, type FunctionToolConfig, type Usage as NativeUsage } from '@strands-agents/sdk';
import { MiddlewareRuntime, recoveryInputSchema, recoveryToolDescription, type Candidate, type RetrieveArgs, type Scope, type Usage } from '@caveman-ai/sdk/middleware';
import { currentOwner, manifest, observe, withOwner, type Attempt } from './common.js';
import { adapterCompatible, frameworkVersion } from './compatibility.js';

export type StrandsScope = Scope | ((agent:LocalAgent) => Scope);
export interface StrandsOptions { runtime:MiddlewareRuntime; scope:StrandsScope }
const adapter={id:'strands',version:'0.1.0',framework_version:frameworkVersion('@strands-agents/sdk')??'unknown',serialization_revision:'strands-message-v1'};
const count=(n:unknown):number|null=>typeof n==='number'&&Number.isSafeInteger(n)&&n>=0?n:null;
function nativeUsage(value:NativeUsage|undefined):Usage|null{
  if(!value)return null;
  const input=count(value.inputTokens),output=count(value.outputTokens);
  return {provenance:'client_observed_sdk',complete:input!==null&&output!==null,input_tokens:input,output_tokens:output,
    cache_read_tokens:count(value.cacheReadInputTokens),cache_write_tokens:count(value.cacheWriteInputTokens),reasoning_tokens:null};
}

/** Native streamAggregated and agent orchestration remain in Strands. */
export class CavemanStrandsModel<T extends BaseModelConfig=BaseModelConfig> extends Model<T>{
  registration:StrandsRegistration|null=null;
  private readonly versionSupported:boolean;
  constructor(readonly inner:Model<T>,readonly options:StrandsOptions){
    super();this.versionSupported=adapterCompatible('strands');
    if(!this.versionSupported&&options.runtime.mode!=='off')options.runtime.decline('unsupported_version');
  }
  override get stateful(){return this.inner.stateful;}
  updateConfig(config:T){this.inner.updateConfig(config);}
  getConfig():T{return this.inner.getConfig();}
  override countTokens(messages:Message[],options?:CountTokensOptions){return this.inner.countTokens(messages,options);}
  private scope():Scope{
    if(typeof this.options.scope!=='function')return this.options.scope;
    const agent=this.registration?.agent?.deref();
    if(!agent)throw new Error('A dynamic Strands scope requires the native agent registration');
    return this.options.scope(agent);
  }

  private async prepare(messages:Message[],options:StreamOptions):Promise<{messages:Message[];attempt:Attempt|null}>{
    if(options.cancelSignal?.aborted&&!currentOwner())this.options.runtime.report(null,{reason:'cancelled',adapter:adapter.id});
    options.cancelSignal?.throwIfAborted();
    if(currentOwner())return {messages,attempt:null};
    const passive=(reason:string)=>({messages,attempt:{runtime:this.options.runtime,
      scope:{namespace:'caveman-report',session_id:'passive',branch_id:'main',cache_epoch:'0'},
      logicalCallId:crypto.randomUUID(),attemptId:crypto.randomUUID(),optimization:null,wireSHA256:null,passive:true,reason,adapter:adapter.id}});
    if(this.options.runtime.mode==='off')return passive('disabled');
    if(!this.versionSupported)return passive('unsupported_version');
    if(this.stateful)return passive('opaque_context');
    const context=await manifest([{system:options.systemPrompt??null},...messages.map(m=>m.toJSON())]);
    if(!context)return passive('unsupported_shape');
    const names=new Map<string,string>();
    for(const message of messages)for(const block of message.content)if(block.type==='toolUseBlock')names.set(block.toolUseId,block.name);
    const candidates:Candidate[]=[],paths=new Map<string,{mi:number;bi:number;pi:number}>();
    messages.forEach((message,mi)=>message.content.forEach((block,bi)=>{
      if(block.type!=='toolResultBlock'||block.status==='error'||!names.has(block.toolUseId)||names.get(block.toolUseId)==='caveman_retrieve')return;
      block.content.forEach((part,pi)=>{
        if(part.type!=='textBlock'||Object.getPrototypeOf(part)?.constructor?.name!=='TextBlock')return;
        const id=`message-${mi}.block-${bi}.part-${pi}`;candidates.push({id,sourceId:id,content:part.text});paths.set(id,{mi,bi,pi});
      });
    }));
    const scope=this.scope(),runtime=this.options.runtime;
    const automatic=!options.toolChoice||'auto'in options.toolChoice;
    const binding=automatic&&this.registration?.bound(options)?runtime.recovery(scope):null;
    const attempt:Attempt={runtime,scope,logicalCallId:crypto.randomUUID(),attemptId:crypto.randomUUID(),optimization:null,wireSHA256:null,adapter:adapter.id};
    const result=await runtime.optimize({scope,adapter,manifest:context,candidates,binding,logicalCallId:attempt.logicalCallId,attemptId:attempt.attemptId,
      model:this.inner.modelId?{provider:this.inner.constructor.name,id:this.inner.modelId,protocol:'strands'}:null,
      ...(binding?{recoveryOverheadText:JSON.stringify(this.registration!.recoveryTool.toolSpec)}:{}),...(options.cancelSignal?{signal:options.cancelSignal}:{})});
    attempt.optimization=result.replacements.length?null:result;
    if(!result.replacements.length)return {messages,attempt};
    if(!result.replacements.every(r=>paths.has(r.segment_id))){attempt.reason='invalid_plan';return {messages,attempt};}
    if(binding&&!this.registration?.bound(options)){attempt.reason='recovery_unavailable';return {messages,attempt};}
    const view=messages.slice();
    for(const replacement of result.replacements){
      const {mi,bi,pi}=paths.get(replacement.segment_id)!,message=view[mi]!,block=message.content[bi] as ToolResultBlock;
      const parts=block.content.slice();parts[pi]=new TextBlock(replacement.text);
      const blocks=message.content.slice();blocks[bi]=new ToolResultBlock({...block,content:parts});
      view[mi]=new Message({...message,content:blocks});
    }
    attempt.optimization=result;
    return {messages:view,attempt};
  }

  async *stream(messages:Message[],options:StreamOptions={}):AsyncGenerator<ModelStreamEvent>{
    const prepared=await this.prepare(messages,options),attempt=prepared.attempt;
    const iterator=attempt?withOwner(attempt,()=>this.inner.stream(prepared.messages,options)[Symbol.asyncIterator]()):this.inner.stream(prepared.messages,options)[Symbol.asyncIterator]();
    if(!attempt){yield* {[Symbol.asyncIterator]:()=>iterator};return;}
    observe(attempt,'dispatch_intent');let usage:Usage|null=null,done=false;
    try{
      while(true){
        const next=await withOwner(attempt,()=>iterator.next());
        if(next.done){done=true;observe(attempt,'completed',usage);return;}
        if(!attempt.passive&&next.value.type==='modelMetadataEvent')usage=nativeUsage(next.value.usage)??usage;
        yield next.value;
      }
    }catch(error){done=true;observe(attempt,options.cancelSignal?.aborted?'cancelled':'failed');throw error;}
    finally{if(!done)observe(attempt,'cancelled');await withOwner(attempt,()=>iterator.return?.());}
  }
}

class StrandsRegistration implements Plugin{
  readonly name='caveman:middleware';
  agent:WeakRef<LocalAgent>|null=null;
  readonly recoveryTool:FunctionTool;
  constructor(readonly options:StrandsOptions){
    this.recoveryTool=new FunctionTool({name:'caveman_retrieve',description:recoveryToolDescription,inputSchema:{...structuredClone(recoveryInputSchema),required:[...recoveryInputSchema.required]} as NonNullable<FunctionToolConfig['inputSchema']>,
      callback:async(input,context)=>{
        const scope=typeof options.scope==='function'?options.scope(context.agent):options.scope;
        return JSON.stringify(await options.runtime.retrieve(scope,input as RetrieveArgs,context.cancelSignal));
      }});
  }
  initAgent(agent:LocalAgent){
    if(this.agent&&this.agent.deref()!==agent)throw new Error('Create a separate Caveman Strands bundle for each agent');
    this.agent=new WeakRef(agent);
    // Registration runs after the host discovers its own tools. A collision
    // keeps the host's tool and prevents this model wrapper from using lossiness.
    if(this.options.runtime.mode==='compress'&&!agent.toolRegistry.get('caveman_retrieve'))agent.toolRegistry.add(this.recoveryTool);
  }
  bound(options:StreamOptions):boolean{
    const agent=this.agent?.deref(),matches=options.toolSpecs?.filter(t=>t.name==='caveman_retrieve')??[];
    return agent?.toolRegistry.get('caveman_retrieve')===this.recoveryTool&&matches.length===1&&JSON.stringify(matches[0])===JSON.stringify(this.recoveryTool.toolSpec);
  }
}

export function withCavemanStrandsModel<T extends BaseModelConfig>(model:Model<T>,options:StrandsOptions):Model<T>{
  return new CavemanStrandsModel(model,options);
}

export function withCavemanStrands(input:AgentConfig&{model:Model},options:StrandsOptions):AgentConfig{
  const model=new CavemanStrandsModel(input.model,options);
  if(options.runtime.mode==='off'||!adapterCompatible('strands'))return {...input,model};
  const registration=new StrandsRegistration(options);
  model.registration=registration;
  return {...input,model,plugins:[...(input.plugins??[]),registration]};
}
